From 7c24a7e3181ef57895ec2e448e3244b17afca328 Mon Sep 17 00:00:00 2001
From: Stefan Herbrechtsmeier <stefan@herbrechtsmeier.net>
Date: Wed, 21 May 2014 15:21:55 +0200
Subject: [PATCH 2/5] omap3: overo: Call bootm only after successful nand read

Signed-off-by: Stefan Herbrechtsmeier <stefan@herbrechtsmeier.net>
---
 include/configs/omap3_overo.h | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/include/configs/omap3_overo.h b/include/configs/omap3_overo.h
index b17e495..d042eea 100644
--- a/include/configs/omap3_overo.h
+++ b/include/configs/omap3_overo.h
@@ -131,8 +131,9 @@
 		"bootz ${loadaddr} - ${fdtaddr}\0" \
 	"nandboot=echo Booting from nand ...; " \
 		"run nandargs; " \
-		"nand read ${loadaddr} linux; " \
-		"bootm ${loadaddr}\0" \
+		"if nand read ${loadaddr} linux; then " \
+			"bootm ${loadaddr};" \
+		"fi;\0" \
 
 #define CONFIG_BOOTCOMMAND \
 	"mmc dev ${mmcdev}; if mmc rescan; then " \
-- 
2.1.0

