From 729dd2bcec926a44c623f4d6da1aa05c3fee9a2d Mon Sep 17 00:00:00 2001
From: Stefan Herbrechtsmeier <stefan@herbrechtsmeier.net>
Date: Wed, 21 May 2014 15:30:12 +0200
Subject: [PATCH 4/8] omap3: overo: Change GUMSTIX_NO_EEPROM to differ from
 empty EEPROM

Signed-off-by: Stefan Herbrechtsmeier <stefan@herbrechtsmeier.net>
---
 board/overo/overo.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/board/overo/overo.c b/board/overo/overo.c
index cf227a3..51a7932 100644
--- a/board/overo/overo.c
+++ b/board/overo/overo.c
@@ -52,7 +52,7 @@ DECLARE_GLOBAL_DATA_PTR;
 
 #define ETTUS_USRP_E			0x01000300
 
-#define GUMSTIX_NO_EEPROM		0xffffffff
+#define GUMSTIX_NO_EEPROM		0xffff0000
 
 static struct {
 	unsigned int device_vendor;
-- 
2.0.0

