From 5ebe0f864b95e02803e43da2c01b28667451f146 Mon Sep 17 00:00:00 2001
From: Stefan Herbrechtsmeier <stefan@herbrechtsmeier.net>
Date: Wed, 21 May 2014 15:42:26 +0200
Subject: [PATCH 7/8] omap3: overo: Add support for the AMiRo expansion board

Signed-off-by: Stefan Herbrechtsmeier <stefan@herbrechtsmeier.net>
---
 board/overo/overo.c |   9 +++++
 board/overo/overo.h | 108 ++++++++++++++++++++++++++++++++++++++++++++++++++++
 2 files changed, 117 insertions(+)

diff --git a/board/overo/overo.c b/board/overo/overo.c
index ab45576..4088a6c 100644
--- a/board/overo/overo.c
+++ b/board/overo/overo.c
@@ -52,6 +52,8 @@ DECLARE_GLOBAL_DATA_PTR;
 
 #define ETTUS_USRP_E			0x01000300
 
+#define AMIRO				0x01000400
+
 #define GUMSTIX_NO_EEPROM		0xffff0000
 
 static struct {
@@ -354,6 +356,13 @@ int misc_init_r(void)
 		MUX_USRP_E();
 		setenv("defaultdisplay", "dvi");
 		break;
+	case AMIRO:
+		printf("Recognized AMiRo mini robot expansion board (rev %d %s)\n",
+			expansion_config.revision,
+			expansion_config.fab_revision);
+		MUX_AMIRO();
+		setenv("expansion", "amiro");
+		break;
 	case GUMSTIX_NO_EEPROM:
 		puts("No EEPROM on expansion board\n");
 		break;
diff --git a/board/overo/overo.h b/board/overo/overo.h
index d0edf86..ba5e4d9 100644
--- a/board/overo/overo.h
+++ b/board/overo/overo.h
@@ -377,6 +377,114 @@ const omap3_sysinfo sysinfo = {
 	MUX_VAL(CP(ETK_D8_ES2),		(IEN  | PTU | EN  | M4)) /*GPIO_22*/\
 	MUX_VAL(CP(ETK_D9_ES2),		(IEN  | PTU | EN  | M4)) /*GPIO_23*/\
 
+#define MUX_AMIRO() \
+/*Overo*/\
+ /*Bluetooth*/\
+ 	MUX_VAL(CP(MCBSP3_DX),		(IEN  | PTD | EN  | M4)) /*GPIO_140*/\
+	MUX_VAL(CP(MCBSP3_DR),		(IEN  | PTD | EN  | M4)) /*GPIO_141*/\
+	MUX_VAL(CP(MCBSP3_CLKX),	(IEN  | PTD | EN  | M4)) /*GPIO_142*/\
+	MUX_VAL(CP(MCBSP3_FSX),		(IEN  | PTD | EN  | M4)) /*GPIO_143*/\
+/*AMIRO Expansion*/\
+ /*GPMC*/\
+	MUX_VAL(CP(GPMC_NCS1),		(IDIS | PTU | EN  | M0)) /*GPMC_nCS1*/\
+	MUX_VAL(CP(GPMC_NCS4),		(IEN  | PTU | EN  | M1)) /*SYS_nDMAREQ1*/\
+	MUX_VAL(CP(GPMC_NCS5),		(IDIS | PTU | EN  | M0)) /*GPMC_nCS5*/\
+	MUX_VAL(CP(GPMC_NCS6),		(IEN  | PTD | DIS | M0)) /*GPMC_nCS6*/\
+	MUX_VAL(CP(GPMC_WAIT1),		(IEN  | PTU | EN  | M4)) /*GPIO_63*/\
+	MUX_VAL(CP(GPMC_WAIT2),		(IEN  | PTD | EN  | M4)) /*GPIO_64*/\
+	MUX_VAL(CP(GPMC_WAIT3),		(IEN  | PTD | EN  | M4)) /*GPIO_65*/\
+ /*DSS*/\
+	MUX_VAL(CP(DSS_PCLK),		(IEN  | PTD | DIS | M4)) /*GPIO_66*/\
+								 /* - SYS_PD_N*/\
+	MUX_VAL(CP(DSS_HSYNC),		(IEN  | PTD | DIS | M4)) /*GPIO_67*/\
+								 /* - SYS_UART_UP*/\
+	MUX_VAL(CP(DSS_VSYNC),		(IEN  | PTD | DIS | M4)) /*GPIO_68*/\
+								 /* - SYS_UART_DN*/\
+	MUX_VAL(CP(DSS_ACBIAS),		(IEN  | PTD | DIS | M4)) /*GPIO_69*/\
+								 /* - SYS_SPI_DIR*/\
+	MUX_VAL(CP(DSS_DATA0),		(IEN  | PTD | DIS | M2)) /*UART1_CTS*/\
+								 /* - BT_RTS*/\
+	MUX_VAL(CP(DSS_DATA1),		(IDIS | PTD | DIS | M2)) /*UART1_RTS*/\
+								 /* - BT_CTS*/\
+	MUX_VAL(CP(DSS_DATA2),		(IEN  | PTD | EN  | M4)) /*GPIO_72*/\
+	MUX_VAL(CP(DSS_DATA3),		(IEN  | PTD | EN  | M4)) /*GPIO_73*/\
+	MUX_VAL(CP(DSS_DATA4),		(IEN  | PTD | EN  | M4)) /*GPIO_74*/\
+	MUX_VAL(CP(DSS_DATA5),		(IEN  | PTD | EN  | M4)) /*GPIO_75*/\
+	MUX_VAL(CP(DSS_DATA6),		(IDIS | PTD | DIS | M2)) /*UART1_TX*/\
+								 /* - BT_RX*/\
+	MUX_VAL(CP(DSS_DATA7),		(IEN  | PTD | DIS | M2)) /*UART1_RX*/\
+								 /* - BT_TX*/\
+	MUX_VAL(CP(DSS_DATA8),		(IEN  | PTD | DIS | M4)) /*GPIO_78*/\
+								 /* - SYS_PROG_N*/\
+	MUX_VAL(CP(DSS_DATA9),		(IEN  | PTD | DIS | M4)) /*GPIO_79*/\
+								 /* - SYS_DONE*/\
+	MUX_VAL(CP(DSS_DATA10),		(IDIS | PTD | DIS | M4)) /*GPIO_80*/\
+								 /* - AUDIO_GAIN0*/\
+	MUX_VAL(CP(DSS_DATA11),		(IDIS | PTD | DIS | M4)) /*GPIO_81*/\
+								 /* - AUDIO_GAIN1*/\
+	MUX_VAL(CP(DSS_DATA12),		(IDIS | PTD | DIS | M4)) /*GPIO_82*/\
+								 /* - BT_EN*/\
+	MUX_VAL(CP(DSS_DATA13),		(IEN  | PTD | EN  | M4)) /*GPIO_83*/\
+	MUX_VAL(CP(DSS_DATA14),		(IEN  | PTD | EN  | M4)) /*GPIO_84*/\
+	MUX_VAL(CP(DSS_DATA15),		(IEN  | PTD | DIS | M4)) /*GPIO_85*/\
+								 /* - CAN_INT_N*/\
+	MUX_VAL(CP(DSS_DATA16),		(IEN  | PTD | EN  | M4)) /*GPIO_86*/\
+	MUX_VAL(CP(DSS_DATA17),		(IDIS | PTD | DIS | M4)) /*GPIO_87*/\
+								 /* - CAN_RST_N*/\
+	MUX_VAL(CP(DSS_DATA18),		(IEN  | PTD | DIS | M2)) /*SPI3_CLK*/\
+	MUX_VAL(CP(DSS_DATA19),		(IEN  | PTD | DIS | M2)) /*SPI3_SIMO*/\
+	MUX_VAL(CP(DSS_DATA20),		(IEN  | PTD | DIS | M2)) /*SPI3_SOMI*/\
+	MUX_VAL(CP(DSS_DATA21),		(IEN  | PTD | EN  | M2)) /*SPI3_CS0*/\
+	MUX_VAL(CP(DSS_DATA22),		(IDIS | PTD | DIS | M4)) /*GPIO_92*/\
+								 /* - AUDIO_SD_N*/\
+	MUX_VAL(CP(DSS_DATA23),		(IDIS | PTD | DIS | M4)) /*GPIO_93*/\
+								 /* - WLAN_EN*/\
+ /*CAMERA*/\
+	MUX_VAL(CP(CAM_FLD),		(IDIS | PTD | DIS | M4)) /*GPIO_98*/\
+								 /* - SYS_CIF_PD_N*/\
+	MUX_VAL(CP(CAM_XCLKB),		(IDIS | PTD | DIS | M0)) /*CAM_XCLKB*/\
+	MUX_VAL(CP(CAM_WEN),		(IDIS | PTD | DIS | M2)) /*CAM_SHUTTER*/\
+	MUX_VAL(CP(CAM_STROBE),		(IDIS | PTD | DIS | M0)) /*CAM_STROBE*/\
+	MUX_VAL(CP(CSI2_DX1),		(IEN  | PTD | DIS | M4)) /*GPIO_114*/\
+								 /* - SYS_CIF_RST_N*/\
+ /*Bluetooth*/\
+	MUX_VAL(CP(UART2_CTS),		(IDIS | PTD | DIS | M4)) /*GPIO_144*/\
+								 /* - LED_GREEN*/\
+	MUX_VAL(CP(UART2_RTS),		(IEN  | PTD | DIS | M4)) /*GPIO_145*/\
+	MUX_VAL(CP(UART2_TX),		(IEN  | PTD | EN  | M4)) /*GPIO_146*/\
+								 /* - SYS_UART_RX*/\
+	MUX_VAL(CP(UART2_RX),		(IEN  | PTD | EN  | M4)) /*GPIO_147*/\
+								 /* - SYS_UART_TX*/\
+	MUX_VAL(CP(UART1_TX),		(IEN  | PTD | EN  | M4)) /*GPIO_148*/\
+	MUX_VAL(CP(UART1_CTS),		(IEN  | PTD | EN  | M4)) /*GPIO_150*/\
+	MUX_VAL(CP(UART1_RX),		(IEN  | PTD | EN  | M4)) /*GPIO_151*/\
+ /*Serial Interface*/\
+	MUX_VAL(CP(UART3_CTS_RCTX),	(IEN  | PTD | DIS | M4)) /*GPIO_163*/\
+								 /* - SYS_UART_CK*/\
+	MUX_VAL(CP(HDQ_SIO),		(IEN  | PTU | DIS | M4)) /*GPIO_170*/\
+								 /* - WLAN_IRQ*/\
+	MUX_VAL(CP(MCSPI1_CLK),		(IEN  | PTD | DIS | M0)) /*McSPI1_CLK*/\
+	MUX_VAL(CP(MCSPI1_SIMO),	(IEN  | PTD | DIS | M0)) /*McSPI1_SIMO*/\
+	MUX_VAL(CP(MCSPI1_SOMI),	(IEN  | PTD | DIS | M0)) /*McSPI1_SOMI*/\
+	MUX_VAL(CP(MCSPI1_CS0),		(IEN  | PTU | EN  | M0)) /*McSPI1_CS0*/\
+	MUX_VAL(CP(MCSPI1_CS1),		(IEN  | PTU | EN  | M0)) /*McSPI1_CS1*/\
+	MUX_VAL(CP(MCSPI1_CS2),		(IEN  | PTU | EN  | M4)) /*GPIO_176*/\
+								 /* - SYS_SPI_INT1_N*/\
+ /*Control and debug */\
+	MUX_VAL(CP(SYS_CLKOUT1),	(IEN  | PTU | EN  | M4)) /*GPIO_10*/\
+	MUX_VAL(CP(SYS_CLKOUT2),	(IEN  | PTU | EN  | M4)) /*GPIO_186*/\
+	MUX_VAL(CP(ETK_CLK_ES2),	(IEN  | PTU | EN  | M2)) /*MMC3_CLK*/\
+	MUX_VAL(CP(ETK_CTL_ES2),	(IEN  | PTU | EN  | M2)) /*MMC3_CMD*/\
+	MUX_VAL(CP(ETK_D0_ES2),		(IEN  | PTU | EN  | M4)) /*GPIO_14*/\
+								 /* - SYS_SPI_INT0_N*/\
+	MUX_VAL(CP(ETK_D3_ES2),		(IEN  | PTU | EN  | M2)) /*MMC3_DAT3*/\
+	MUX_VAL(CP(ETK_D4_ES2),		(IEN  | PTU | EN  | M2)) /*MMC3_DAT0*/\
+	MUX_VAL(CP(ETK_D5_ES2),		(IEN  | PTU | EN  | M2)) /*MMC3_DAT1*/\
+	MUX_VAL(CP(ETK_D6_ES2),		(IEN  | PTU | EN  | M2)) /*MMC3_DAT2*/\
+	MUX_VAL(CP(ETK_D7_ES2),		(IEN  | PTU | EN  | M4)) /*GPIO_21*/\
+	MUX_VAL(CP(ETK_D8_ES2),		(IEN  | PTU | EN  | M4)) /*GPIO_22*/\
+	MUX_VAL(CP(ETK_D9_ES2),		(IEN  | PTU | EN  | M4)) /*GPIO_23*/
+
 #define MUX_OVERO_SDIO2_DIRECT() \
 	MUX_VAL(CP(MMC2_CLK),		(IEN  | PTU | EN  | M0)) /*MMC2_CLK*/\
 	MUX_VAL(CP(MMC2_CMD),		(IEN  | PTU | EN  | M0)) /*MMC2_CMD*/\
-- 
2.0.0

