From 261854ac0dcaa370559fc9839cbe3b08c4b6b706 Mon Sep 17 00:00:00 2001
From: Timo Korthals <tkorthals@techfak.uni-bielefeld.de>
Date: Fri, 30 Jan 2015 20:44:20 +0100
Subject: [PATCH] Start OMAP watchdog with probing

---
 drivers/watchdog/omap_wdt.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/watchdog/omap_wdt.c b/drivers/watchdog/omap_wdt.c
index 09cf013..abaaeb6 100644
--- a/drivers/watchdog/omap_wdt.c
+++ b/drivers/watchdog/omap_wdt.c
@@ -266,8 +266,6 @@ static int omap_wdt_probe(struct platform_device *pdev)
 	omap_wdt->bootstatus = (rs & (1 << OMAP_MPU_WD_RST_SRC_ID_SHIFT)) ?
 				WDIOF_CARDRESET : 0;
 
-	omap_wdt_disable(wdev);
-
 	ret = watchdog_register_device(omap_wdt);
 	if (ret) {
 		pm_runtime_disable(wdev->dev);
@@ -280,6 +278,8 @@ static int omap_wdt_probe(struct platform_device *pdev)
 
 	pm_runtime_put_sync(wdev->dev);
 
+        omap_wdt_start(omap_wdt);
+
 	return 0;
 }
 
-- 
2.2.2

