diff -aur a/arch/arm/mach-omap2/board-amiro.c b/arch/arm/mach-omap2/board-amiro.c
--- a/arch/arm/mach-omap2/board-amiro.c	2013-07-22 22:30:13.000000000 +0200
+++ b/arch/arm/mach-omap2/board-amiro.c	2013-07-22 22:34:23.000000000 +0200
@@ -34,6 +35,8 @@
 #include <linux/regulator/machine.h>
 #include <linux/regulator/fixed.h>
 #include <linux/spi/spi.h>
+#include <linux/i2c.h>
+#include <linux/i2c/at24.h>
 #include <linux/irq.h>
 
 #include <linux/mtd/mtd.h>
@@ -223,10 +226,17 @@
 /*
  * I2C
  */
-static struct i2c_board_info __initdata amiro_i2c3_info[] = {
-       {
-               I2C_BOARD_INFO("eeprom", 0x51),
-       },
+static struct at24_platform_data m24c01_config = {
+	.byte_len	= SZ_1K / 8,
+	.page_size	= 8,
+};
+
+static struct i2c_board_info amiro_i2c3_board_info[] __initdata = {
+	{
+		/* Configuration EEPROM */
+		I2C_BOARD_INFO("at24" /* 24c01 */, 0x51u),
+		.platform_data = &m24c01_config,
+	},
 };
 
 static int __init amiro_i2c_init(void)
@@ -236,9 +246,12 @@
 			TWL_COMMON_REGULATOR_VPLL2);
 
 	omap3_pmic_init("tps65950", &amiro_twl_pdata);
-	omap_register_i2c_bus(3, 100, amiro_i2c3_info,
-			ARRAY_SIZE(amiro_i2c3_info));
-
+	omap_register_i2c_bus(
+	                      3, /* bus num */
+	                      400, /* freq kHz */
+	                      amiro_i2c3_board_info,
+	                      ARRAY_SIZE(amiro_i2c3_board_info)
+	                     );
 	return 0;
 }
 
